## PHYS 6260: Homework 5

In [1]:
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt

## Problem 1: A circuit of resistors

![alt text](pb105.JPG "Circuit")

Consider the above circuit. All of the resistors have the same resistance $R$. The power rail at the top is at voltage $V_+ = 5$ V. To find the other four voltages marked in the circuit diagram, we use Ohm's law and the Kirchhoff junction law, which says that the total current into any junction must equal the total current out of the same junction, $\sum_{i} I_{in,i}=\sum_{i} I_{out,i}$. Thus, for the junction at voltage V_1, we have
$$\frac{V_1-V_2}R + \frac{V_1-V_3}R+\frac{V_1-V_4}R+\frac{V_1-V_+}R = 0$$
or equivalently
$$4V_1-V_2-V_3-V_4 =V_+$$

### Part (a)

Similarly, we found equations for the other three junctions with unknown voltages:
$$\frac{V_2}R + \frac{V_2-V_4}R+\frac{V_2-V_1}R= 0 \rightarrow 3V_2-V_1-V_4 =0$$
$$\frac{V_3-V_1}R + \frac{V_3-V_4}R+\frac{V_3-V_+}R= 0 \rightarrow 3V_3-V_1-V_4=V_+$$
$$\frac{V_4}R + \frac{V_4-V_2}R+\frac{V_4-V_1}R+\frac{V_4-V_3}R= 0 \rightarrow 4V_4-V_1-V_2 -V_3 =0$$

### Part (b)

Using Gaussian elimination with backsubstition, we found the four voltages:

In [2]:
#initial the matrix and vector for Ax = v with x = (V1,V2,V3,V4)
A = np.array([[4,-1,-1,-1],
              [-1,3,0,-1],
              [-1,0,3,-1],
              [-1,-1,-1,4]], float)
v = np.array([5,0,5,0], float)
N = len(v)

#using the code writing in class as a basis:
#Gaussian elimination
for m in range(N):
    #divide by the diagonal element
    diagonal = A[m,m]
    A[m,:] = A[m,:] / diagonal
    v[m] = v[m] / diagonal
    # Now subtract from the lower rows 
    for i in range(m+1, N):
        mul = A[i,m]
        A[i,:] -= mul * A[m,:]
        v[i] -= mul * v[m]
# Backsubstitution 
x = np.empty(N, float)
for m in range(N-1, -1, -1): 
    x[m] = v[m]
    for i in range(m+1, N):
        x[m] -= A[m, i]*x[i]
    
print("The four voltages are: V_1 = %.2f V, V_2 = %.2f V, V_3 = %.2f V, V_4 = %.2f V." %(x[0],x[1],x[2],x[3]))

The four voltages are: V_1 = 3.00 V, V_2 = 1.67 V, V_3 = 3.33 V, V_4 = 2.00 V.


<font color='red'>Correct 8/8</font>

## Problem 2: Asymmetric quantum well

Quantum mechanics can be formlated as matrix problem and solves on a computer using linear algebra methods. Suppose, for example, we have a particle of mass $M$ in a 1D quantum well of width $L$, but not a square well like the examples you've probably seen before. Instead consider an arbitrary potential $V(x)$ that varies inside the well, like what's shown below. We cannot solve such problem analytically in general, but we can solve them using computational methods.
![alt text](pb205.JPG "Circuit")

In a pure state of energy E, the spatial part of the wave function obeys the time-independent Schrodinger equation $\hat{H}$ is given by
$$\hat{H} = -\frac{\hbar^2}{2M} \frac{d^2}{dx^2} + V(x)$$

For simplicity, let's assume that the walls of the well are infinitely high, so that the wave function is zero outside the well, which means it must go to zero at $x=0$ and $z=L$. In that case, the wave function can be expressed as a Fourier sine series as 

$$\psi(x) = \sum_{n=1}^{\infty}\psi_{n}\sin{\frac{\pi n x}{L}},$$

where $\psi_1, \psi_2,...$ are the Fourier coefficients.

### Part (a)

$$\hat{H}\psi=E\psi\rightarrow \hat{H}\sum_{n=1}^{\infty}\psi_{n}\sin{\frac{\pi n x}{L}}=E \sum_{n=1}^{\infty}\psi_{n}\sin{\frac{\pi n x}{L}} $$
Multiple both sides by $\sin{\frac{\pi mx}{L}}$ and integrate from 0 to L:
$$\int_{0}^{L}\hat{H}\sum_{n=1}^{\infty}\psi_{n}\sin{\frac{\pi n x}{L}}\sin{\frac{\pi m x}{L}}dx = \int_{0}^{L}E\sum_{n=1}^{\infty}\psi_{n}\sin{\frac{\pi n x}{L}}\sin{\frac{\pi m x}{L}}dx$$ 
Or $$\sum_{n=1}^{\infty}\psi_{n}\int_{0}^{L}\sin{\frac{\pi m x}{L}}\hat{H}\sin{\frac{\pi n x}{L}}dx = E\sum_{n=1}^{\infty}\psi_{n}\int_{0}^{L}\sin{\frac{\pi n x}{L}}\sin{\frac{\pi m x}{L}}dx$$ 
and
$$\int_{0}^{L}\sin{\frac{\pi m x}{L}}\sin{\frac{\pi n x}{L}} dx
= 
\begin{cases}
\frac{L}{2} & \text{if}\ m =n, \\
0 & \text{otherwise}
\end{cases}$$ 
So, on the right hand side, everything vanishes except m = n. Thus,
$$\sum_{n=1}^{\infty}\psi_{n}\int_{0}^{L}\sin{\frac{\pi m x}{L}}\hat{H}\sin{\frac{\pi n x}{L}}dx = \frac{1}{2}LE\psi_{m}$$ 
$$\rightarrow \sum_{n=1}^{\infty}\psi_{n}\frac{2}{L}\int_{0}^{L}\sin{\frac{\pi m x}{L}}\hat{H}\sin{\frac{\pi n x}{L}}dx = E\psi_{m}$$
Hence, defining a matrix **H** with elements
$$H_{mn} = \frac{2}{L}\int_{0}^{L}\sin \frac{\pi m x}{L}\hat{H}\sin\frac{\pi n x}{L}dx 
= \frac{2}{L}\int_{0}^{L}\sin \frac{\pi m x}{L}\bigg[-\frac{\hbar^2}{2M}\frac{d^2}{dx^2} + V(x)\bigg]\sin\frac{\pi n x}{L}dx$$
The Schrodinger's equation can be written in matrix form as **H**$\psi = E\psi$, where $\psi$ is the vector ($\psi_1, \psi_2,...$). Thus $\psi$ is an eigenvector of the *Hamiltonian matrix* **H** with an eigenvalue *E*. If we calculate the eigenvalues of this matrix, then we know the allowed energies of the particle in the well.

### Part (b)

For the case $V(x) = ax/L$:
$$H_{mn} = \frac{2}{L}\int_{0}^{L}\sin\frac{\pi m x}{L}\bigg[-\frac{\hbar^2}{2M}\frac{d^2}{dx^2} + \frac{ax}{L}\bigg]\sin\frac{\pi n x}{L}dx = \frac{2}{L}\int_{0}^{L}\sin\frac{\pi m x}{L}\bigg[-\frac{\hbar^2}{2M}\frac{d^2}{dx^2} \bigg]\sin\frac{\pi n x}{L}dx +\frac{2}{L}\int_{0}^{L}\sin\frac{\pi m x}{L}\frac{ax}{L}\sin\frac{\pi n x}{L}dx$$
but 
$$-\frac{\hbar^2}{2M}\frac{d^2}{dx^2}\sin\frac{\pi nx}{L}=-\frac{\hbar^2}{2M}\frac{d}{dx} \cos\frac{\pi nx}{L}\frac{\pi n}{L}=\frac{\hbar^2}{2M}\sin\frac{\pi nx}{L}\frac{\pi^2 n^2}{L^2}$$
So, 
$$H_{mn}=\frac{\hbar^2\pi^2 n^2 }{ML^3} \int_{0}^{L}\sin\frac{\pi m x}{L}\sin\frac{\pi n x}{L}dx + \frac{2a}{L^2}\int_{0}^{L}x \sin\frac{\pi m x}{L}\sin\frac{\pi n x}{L}dx $$

Using the given integral results: 

$$\int_{0}^{L}\sin{\frac{\pi m x}{L}}\sin{\frac{\pi n x}{L}} dx
= 
\begin{cases}
\frac{L}{2} & \text{if}\ m =n, \\
0 & \text{otherwise}
\end{cases}$$ 
and
$$\int_{0}^{L}x\sin{\frac{\pi mx}{L}}\sin{\frac{\pi nx}{L}}dx 
= 
\begin{cases}
0 & \text{if}\ m \neq n, \text{and}\ m , n \text{ are both even or both odd}\\
-\big(\frac{2L}{\pi}\big)^2 \frac{mn}{(m^2 - n^2)^2} & \text{if}\ m \neq n \text{ and one is even, one is odd,} \\
{L^2}/4 & \text{if}\ m = n 
\end{cases}
$$ 
We found a general expression for the matrix element $H_{mn}$ as
$$H_{mn}=
\begin{cases}
0 & \text{if}\ m \neq n, \text{and}\ m , n \text{ are both even or both odd}\\
-\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & \text{if}\ m \neq n \text{ and one is even, one is odd,} \\
\frac{\hbar^2\pi^2 n^2 }{2ML^2} +\frac{a}{2}& \text{if}\ m = n 
\end{cases}
$$ 
So, the matrix is real and symmetric. This can be showed by an example of 4x4 matrix:
\begin{equation}
H_{mn} =
\begin{bmatrix}
    \frac{\hbar^2\pi^2 n^2 }{2ML^2} +\frac{a}{2} & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & 0 & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2}\\
    -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & \frac{\hbar^2\pi^2 n^2 }{2ML^2} +\frac{a}{2} & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & 0\\
    0 & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & \frac{\hbar^2\pi^2 n^2 }{2ML^2} +\frac{a}{2} & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2}\\
    -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & 0 & -\frac{8a}{\pi^2}\frac{mn}{(m^2 - n^2)^2} & \frac{\hbar^2\pi^2 n^2 }{2ML^2} +\frac{a}{2}\\
\end{bmatrix}
\end{equation}

In [3]:
# constants 
hbar = 6.5821e-16 #eVs
L = 5e-10 #m
a = 10 #eV
M = 9.1094e-31 #kg
e = 1.6022e-19 #C

In [4]:
# evaluate H_mn for arbitrary m and n
def h(m,n):
    if m == n:
        return e*(hbar*np.pi*n)**2/(2*M*L**2) + a/2
    elif not(m%2 == n%2):
        return -(8*m*n*a) / ((np.pi**2)*(m**2-n**2)**2)
    else:
        return 0

### Part (c)

In [5]:
N = 10
H = np.empty([N,N])
for m in range(N):
    for n in range(N):
        H[m, n] = h(m+1,n+1)

eigenv, eigenvec = np.linalg.eigh(H)
print("The first ten energy levels of the quantum well are:")
print(*np.round(eigenv,3), sep="eV, ", end="eV.")

The first ten energy levels of the quantum well are:
5.836eV, 11.181eV, 18.663eV, 29.144eV, 42.655eV, 59.186eV, 78.73eV, 101.286eV, 126.852eV, 155.556eV.

### Part (d)

In [6]:
N1 = 100
H1 = np.empty([N1,N1])
for m1 in range(N1):
    for n1 in range(N1):
        H1[m1, n1] = h(m1+1,n1+1)

eigenv1, eigenvec1 = np.linalg.eigh(H1)
print("The first ten energy levels of the quantum well are:")
print(*np.round(eigenv1[:10],3), sep="eV, ", end="eV.")

The first ten energy levels of the quantum well are:
5.836eV, 11.181eV, 18.663eV, 29.144eV, 42.655eV, 59.186eV, 78.73eV, 101.286eV, 126.851eV, 155.427eV.

Comparing with the values in part (c), the eigenvalues match until n = 9, which in part c is 126.852 compare to 126.851 in part d. And, for n = 10, the eigenvalues is off all fractional part. So, I think the accuracy of this calculation is acceptable but not too high.

<font color='red'>Correct 16/16</font>

## Problem 3: Biochemical equilibrium points

The Biochemical proces of $glycolysis$, the breakdown of glucose in the body to release energy, can be modeled by the equations
$$\frac{dx}{dt}=-x+ay+x^2y, \frac{dy}{dt}=b-ay-x^2y$$ 
Here x and y represent concentrations of two chemicals, ADP and F6P, and a and b are postitive constants. One of the important features of nonlinear equations like these is their $stationary$ $points$, meaning values x and y at which the derivatives of both variables become zero simultaneously, so that the variables stop changing and become constant in time. Setting the derivatives to zero above, the stationary points of our glycolysis equations are the following solutions:
$$-x+ay+x^2y=0, b -ay-x^2y=0$$

### Part (a)

We have 
$$b-ay-x^2y=0 \rightarrow ay+x^2y=b$$ 
and $$-x+ay+x^2y=0 \rightarrow x = ay+x^2y=b$$
$$b-ay-x^2y=0 \rightarrow b-ay-b^2y=0 \rightarrow y(a+b^2)=b \rightarrow y =\frac{b}{a+b^2}$$
Therefore, the solution of the above equations are:
$$x=b, y = \frac{b}{a+b^2}$$

### Part (b)

$$x=ay+x^2y=y(a+x^2)$$
$$x=b \rightarrow y = \frac{b}{a+b^2}=\frac{b}{a+x^2}$$

In [8]:
a = 1
b = 2
accuracy = 1e-6
point = 1000

In [12]:
# using the relaxation method with a = 1 and b = 2
# x = f(x,y)
def f(x, y):
    return y * (a + x ** 2)

# y = g(x,y)
def g(x, y):
    return b / (a + x ** 2)

def points(f, g):
    for i in range(point):
        x1 = 0.1
        y1 = 0.1
        x2 = f(x1, y1)
        y2 = g(x1, y1)
        error1 = 1.0
        error2 = 1.0
        while abs(error1) > accuracy and abs(error2) > accuracy:
            x1, x2, y1, y2 = x2, f(x2, y2), y2, g(x2, y2)
            error1 = abs((x1 - x2) / (2*x2*y2))
            error2 = abs((y1 - y2) / (-2*b*x2 / (a + x2 ** 2)**2))
    return [x2,y2]

points(f,g)

KeyboardInterrupt: 

The method fails to converge to a solution in this case because the absolute value of the derivative is greater than one, which means we're getting farther from the solution on each step. So, it never converges.

### Part (c)

In [13]:
# So, we need to rearrange the equations such that the absolute value of the derivative is less than one.
def f2(x, y):
    return np.sqrt(b / y - a)


def g2(x, y):
    return x / (a + x ** 2)

def points2(f, g):
    for i in range(point):
        x1 = 0.1
        y1 = 0.1
        x2 = f(x1, y1)
        y2 = g(x1, y1)
        error1 = 1.0
        error2 = 1.0
        while abs(error1) > accuracy and abs(error2) > accuracy:
            x1, x2, y1, y2 = x2, f(x2, y2), y2, g(x2, y2)
            error1 = abs((x1 - x2) / (-b/(2*y2**2*np.sqrt(b/y-a))))
            error2 = abs((y1 - y2) / (x / ((a-x**2)/(a + x ** 2)**2)))
    return [x2,y2]

points(f2,g2)

[2.000132986847455, 0.3999841823386661]

In [331]:
x2, y2 = points(f2,g2)
y = b/(a+b**2)
print("The solutions are x2 = %.2f, y2 = %.2f . Compare to part a, which x = %.2f and y = %.2f, the result agrees with part a." %(x2,y2,b,y))

The solutions are x2 = 2.00, y2 = 0.40 . Compare to part a, which x = 2.00 and y = 0.40, the result agrees with part a.


<font color='red'>I'm not really sure why you run two loops here. You're looping through iterations on x and y in the `for` loop for a maximum of 1000 points. That's reasonable. But what's the inner while loop? 6/8</font>

## Problem 4: The Lagrange point

There exists a point between the Earth and the Moon, called the $L_1$ Lagrange point, at which a satellite wil orbit the Earth in perfect synchrony with the Moon, staying always between the two. This works because the inward pull of the Earth and the outward pull of the Moon combine to create exactly the needed centripetal force that keeps the satellite in its orbit, as shown below.
![alt text](pb405.JPG "EM")

### Part (a)

Assuming circular orbits and that the Rarth is much more massive than either the Moon or the satellite. 
Let $M$ and $m$ are the Earth and Moon masses, $G$ is Newton's gravitational constant, and $\omega$ is the angular velocity of both the Moon and satellite. Since the inward pull of the Earth and the outward pull of the Moon combine to create exactly the needed centripetal force that keeps the satellite in its orbit, we have
$$\vec{F_M}+\vec{F_m}=\vec{F}_{centripetal}$$
$$\frac{GMm_S}{r^2}-\frac{Gmm_S}{(R-r)^2}=\frac{v^2m_S}{r}$$
and $v = \omega r$. Thus,
$$\frac{GM}{r^2}-\frac{Gm}{(R-r)^2}=\omega^2r$$

### Part (b)

In [332]:
# Various constants and parameters in the system
G = 6.674e-11 #m^3kg^-1s^-2 
M = 5.974e24 #kg
m = 7.348e22 #kg 
R = 3.844e8 #m 
w = 2.662e-6 #s^-1   

In [333]:
# using Mathematica, I found out the solution is 3.260*10^8 m
# so, my initial values are 2*10^7 and 4*10^9 for the secant method

# To avoid the divided by 0 float from the orginal equation, 
# I get rid of the deniminator by multiple both sides for r^2(R-r)^2
def f(r): 
    return G*M*(R-r)**2 - G*m*r**2 - omega**2*r**3*(R-r)**2

def solve(): 
    r1 = 2.0e7
    r2 = 4.0e9
    delta = 1
    
    while abs(delta)>accuracy:
        delta = f(r2) * (r2-r1)/(f(r2)-f(r1))
        r3 = r2 - f(r2)*(r2-r1)/(f(r2)-f(r1))
        r1 = r2
        r2 = r3
     
    return r3

print("The distance r from Earth to L1 is %.6g m" %solve())

The distance r from Earth to L1 is 3.26045e+08 m


<font color='red'>Correct 8/8</font>